synapse/storage/databases/main/account_data.py

```diff
@@ -210,7 +210,7 @@ def get_account_data_for_room_txn(
             "get_account_data_for_room", get_account_data_for_room_txn
         )
 
-    @cached(num_args=3, max_entries=5000)
+    @cached(num_args=3, max_entries=5000, tree=True)
     async def get_account_data_for_room_and_type(
```

Comment:
This seems OK, where we invalid this it looks like we do something weird with whether there's a `room_id` or not, but I don't think that would change from this PR.